File: Data Structures & Algorithms/invert-a-binary-tree/submission-0.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func invertTree(root *TreeNode) *TreeNode {
	
	var dfs func(node *TreeNode)
	dfs = func(node *TreeNode) {
		if node == nil {
			return
		}
		tmp := node.Left
		node.Left = node.Right
		node.Right = tmp

		dfs(node.Left)
		dfs(node.Right)
	}

	dfs(root)
	return root
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def dfs(node: Optional[TreeNode]) -> (bool, int):
            if not node:
                return True, 0
            
            l_b, l_h = dfs(node.left)
            r_b, r_h = dfs(node.right)

            balanced = (l_b and r_b and abs(l_h - r_h) <= 1)

            return balanced, max(l_h, r_h) + 1
        b, _ = dfs(root)
        return b 